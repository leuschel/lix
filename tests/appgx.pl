/*Generated by Lix*/
:- dynamic flag/2, memo_table/2.
:- use_module(library(lists)).
:- use_module(library(terms)).
:- op(1150, fx, type).
/*oldvalue__1(_7546,_7547)=oldvalue(_7546,_7547)*/
oldvalue__1(A, B) :-
        flag(gensym(A), B), !.
oldvalue__1(_, 0).
/*set_flag__1(_8031,_8010)=set_flag(gensym(_8031),_8010)*/
set_flag__1(A, B) :-
        retract(flag(gensym(A),_)), !,
        asserta(flag(gensym(A),B)).
set_flag__1(A, B) :-
        asserta(flag(gensym(A),B)).
/*gensym__1(_7283)=gensym(app,_7283)*/
gensym__1(A) :-
        var(A),
        oldvalue__1(app, B),
        C is B+1,
        set_flag__1(app, C),
        name(C, D),
        name(A, [97,112,112,95,95|D]).
/*unfold__1(_7810,_7812,_7814,_7785)=unfold(app(_7810,_7812,_7814),_7785)*/
unfold__1([], A, A, true).
unfold__1([A|B], C, [A|D], E) :-
        memo__1(B, C, D, E).
/*flatten__1(_8588,_8589)=flatten(_8588,_8589)*/
flatten__1((A:-B), (A:-C)) :- !,
        flatten__1(B, C).
flatten__1((A,B), C) :- !,
        flatten__1(A, D),
        flatten__1(B, E),
        (   D=true ->
            C=E
        ;   E=true ->
            C=D
        ;   C=(D,E)
        ).
flatten__1((A;B), C) :- !,
        flatten__1(A, D),
        flatten__1(B, E),
        (   D=true ->
            C=E
        ;   E=true ->
            C=D
        ;   C=(D;E)
        ).
flatten__1((A->B;C), (D->E;F)) :- !,
        flatten__1(A, D),
        flatten__1(B, E),
        flatten__1(C, F).
flatten__1(A, A).
/*pretty_print_clauses__1(_8284)=pretty_print_clauses(_8284)*/
pretty_print_clauses__1([]).
pretty_print_clauses__1([A|B]) :-
        flatten__1(A, C),
        portray_clause(C),
        pretty_print_clauses__1(B).
/*memo__1(_6855,_6857,_6859,_6830)=memo(app(_6855,_6857,_6859),_6830)*/
memo__1(A, B, C, D) :-
        (   memo_table(app(A,B,C), D) ->
            true
        ;   gensym__1(E),
            F=..[E,G,H],
            assert(memo_table(app(G,B,H),F)),
            findall((F:-I), unfold__1(G,B,H,I), J),
            format('/*~k=~k*/~n', [F,app(G,B,H)]),
            pretty_print_clauses__1(J),
            memo_table(app(A,B,C), D)
        ).
/*lix__1(_6599,_6601,_6603,_6574)=lix(app(_6599,_6601,_6603),_6574)*/
lix__1(A, B, C, D) :-
        retractall(memo_table(_,_)),
        print('/*Generated by Lix*/\n'),
        print(':- dynamic flag/2, memo_table/2.\n'),
        print(':- use_module(library(lists)).\n'),
        print(':- use_module(library(terms)).\n'),
        print(':- op(1150, fx, type).\n'),
        memo__1(A, B, C, D).
lix(app([a,b,c,d],[b],A), B) :-
        lix__1([a,b,c,d], [b], A, B).
