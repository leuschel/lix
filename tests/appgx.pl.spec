/*Generated by Lix*/
:- dynamic flag/2, memo_table/2.
:- use_module(library(lists)).
:- use_module(library(terms)).
:- op(1150, fx, type).
/*app__1(_6593,_6595)=app(_6593,'.'(b,[]),_6595)*/
app__1([], [b]).
app__1([A|B], [A|C]) :-
        app__1(B, C).
app([a,b,c,d], [b], A) :-
        app__1([a,b,c,d], A).
