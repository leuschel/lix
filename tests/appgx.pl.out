[[a,b,c,d,b]]
